unified-scheduler-pool/src/lib.rs

```diff
@@ -0,0 +1,749 @@
+    next_scheduler_id: AtomicSchedulerId,
+    _phantom: PhantomData<(S, TH)>,
+}
+
+pub type DefaultSchedulerPool =
+    SchedulerPool<PooledScheduler<DefaultTaskHandler>, DefaultTaskHandler>;
+
+impl<S: SpawnableScheduler<TH>, TH: TaskHandler> SchedulerPool<S, TH> {
+    fn new(
+        log_messages_bytes_limit: Option<usize>,
+        transaction_status_sender: Option<TransactionStatusSender>,
+        replay_vote_sender: Option<ReplayVoteSender>,
+        prioritization_fee_cache: Arc<PrioritizationFeeCache>,
+    ) -> Arc<Self> {
+        Arc::new_cyclic(|weak_self| Self {
+            schedulers: Mutex::default(),
+            log_messages_bytes_limit,
+            transaction_status_sender,
+            replay_vote_sender,
+            prioritization_fee_cache,
+            weak_self: weak_self.clone(),
+            next_scheduler_id: AtomicSchedulerId::new(PRIMARY_SCHEDULER_ID),
+            _phantom: PhantomData,
+        })
+    }
+
+    pub fn new_dyn(
+        log_messages_bytes_limit: Option<usize>,
+        transaction_status_sender: Option<TransactionStatusSender>,
+        replay_vote_sender: Option<ReplayVoteSender>,
+        prioritization_fee_cache: Arc<PrioritizationFeeCache>,
+    ) -> InstalledSchedulerPoolArc {
+        Self::new(
+            log_messages_bytes_limit,
+            transaction_status_sender,
+            replay_vote_sender,
+            prioritization_fee_cache,
+        )
+    }
+
+    // See a comment at the weak_self field for justification of this.
+    fn self_arc(&self) -> Arc<Self> {
+        self.weak_self
+            .upgrade()
+            .expect("self-referencing Arc-ed pool")
+    }
+
+    fn new_scheduler_id(&self) -> SchedulerId {
+        self.next_scheduler_id.fetch_add(1, Relaxed)
+    }
+
+    fn return_scheduler(&self, scheduler: Box<S>) {
+        assert!(!scheduler.has_context());
+
+        self.schedulers
+            .lock()
+            .expect("not poisoned")
+            .push(scheduler);
+    }
+
+    fn do_take_scheduler(&self, context: SchedulingContext) -> Box<S> {
+        // pop is intentional for filo, expecting relatively warmed-up scheduler due to having been
+        // returned recently
+        if let Some(mut scheduler) = self.schedulers.lock().expect("not poisoned").pop() {
+            scheduler.replace_context(context);
+            scheduler
+        } else {
+            Box::new(S::spawn(
+                self.self_arc(),
+                context,
+                TH::recreate_handler_with_pool(self),
+            ))
+        }
+    }
+}
+
+impl<S: SpawnableScheduler<TH>, TH: TaskHandler> InstalledSchedulerPool for SchedulerPool<S, TH> {
+    fn take_scheduler(&self, context: SchedulingContext) -> Box<dyn InstalledScheduler> {
+        self.do_take_scheduler(context)
+    }
+}
+
+pub trait TaskHandler: Send + Sync + Debug + Sized + 'static {
+    fn recreate_handler_with_pool<S: SpawnableScheduler<Self>>(
+        pool: &SchedulerPool<S, Self>,
```

Comment:
I couldn't find this parameter being used in your impl branch, but names have changed so I might have missed it.
Is this necessary? It seems odd for a **task** handler, to need knowledge of the pool it is in.

Reply:
no, this isn't strictly needed. so, i just removed it from this pr: f29cabb96ae50d970832175a7c2ce721b71ab478

iirc, this is needed for some `bench`es along side the whole `SEA: ScheduleExecutionArg` mess, which i'm just keeping to postpone for code-review. ;)